### PR TITLE
Feature/issue 320 prob funs

### PR DESCRIPTION
**Pull request by syclik**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
This pull request deprecates the `_log` functions by calling the new `_lpdf` / `_lpmf` / `_lcdf` / `_lccdf` function.

#### How to Verify:
Look at code and run code.

#### Side Effects:
None.

#### Documentation:
Doxygen doc deprecating the `_log` function.

#### Reviewer Suggestions: 
Anyone.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by syclik** [comment]
Just waiting for a review now. This pull request makes the `_log()` function dispatch to the appropriate `_lpdf`, `_lpmf`, `_lcdf`, `_lccdf` function.

File: stan/math/prim/mat/prob/multi_gp_log.hpp [hunk 1]
```diff
@@ -43,61 +33,20 @@ namespace stan {
...
+     */
     template <typename T_y, typename T_covar, typename T_w>
     inline
     typename boost::math::tools::promote_args<T_y, T_covar, T_w>::type
     multi_gp_log(const Eigen::Matrix<T_y, Eigen::Dynamic, Eigen::Dynamic>& y,
                  const Eigen::Matrix<T_covar, Eigen::Dynamic, Eigen::Dynamic>&
                  Sigma,
                  const Eigen::Matrix<T_w, Eigen::Dynamic, 1>& w) {
-      return multi_gp_log<false>(y, Sigma, w);
+      return multi_gp_lpdf<T_y, T_covar, T_w>(y, Sigma, w);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/prob/multi_gp_log.hpp]: For future reference (not this pull request), those types would've all been inferred without listing them explicitly.

File: stan/math/prim/scal/prob/cauchy_log.hpp [hunk 2]
```diff
@@ -1,22 +1,7 @@
 #ifndef STAN_MATH_PRIM_SCAL_PROB_CAUCHY_LOG_HPP
 #define STAN_MATH_PRIM_SCAL_PROB_CAUCHY_LOG_HPP
 
-#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
-#include <stan/math/prim/scal/meta/partials_return_type.hpp>
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/cauchy_log.hpp]: Don't we want to include what we explicitly use in the classes even if it's included indirectly through another header.  Your call.  I'm going to just approve this and let you sort out what you want.

File: stan/math/prim/scal/prob/pareto_log.hpp [hunk 3]
```diff
@@ -1,123 +1,29 @@
...
-#include <boost/random/variate_generator.hpp>
-#include <cmath>
+#include <stan/math/prim/scal/prob/pareto_lpdf.hpp>
 
 namespace stan {
   namespace math {
 
-    // Pareto(y|y_m, alpha)  [y > y_m;  y_m > 0;  alpha > 0]
+    /**
+     * @deprecated use <code>pareto_lpdf</code>
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/pareto_log.hpp]: These are all missing argument and template parameter doc, but I suppose that's OK given that it's deprecated.

**Comment by bob-carpenter** [comment]
I'll leave it up to you whether you want to fix the 

* include-what-you-use problem (return template metaprogram not included),

* the overspecification of types (not wrong, just redundant), and

* missing doc for deprecated functions.

File: stan/math/prim/scal/prob/cauchy_log.hpp (hunk 2 above)
Comment (syclik) [inline stan/math/prim/scal/prob/cauchy_log.hpp]: Yes, absolutely. I'll go through and fix.

File: stan/math/prim/mat/prob/multi_gp_log.hpp (hunk 1 above)
Comment (syclik) [inline stan/math/prim/mat/prob/multi_gp_log.hpp]: I will fix these too.

**Comment by bob-carpenter** [comment]
@syclik  Please let me know if this is ready for review.

**Comment by syclik** [comment]
@bob-carpenter, hold off for a bit. I want to update the prob functions so that they have the right includes; it'll just take a bit of time. Have to touch ~50 files, but it was something I wasn't careful about to start and I want it fixed before it gets in.

**Comment by syclik** [comment]
I just updated. @bob-carpenter or @seantalts or anyone else, want to look through this for a sanity check?

File: stan/math/prim/mat/prob/multi_gp_log.hpp (hunk 1 above)
Comment (syclik) [inline stan/math/prim/mat/prob/multi_gp_log.hpp]: You're absolutely right. I didn't fix these on this pull request.

File: stan/math/prim/scal/prob/cauchy_log.hpp (hunk 2 above)
Comment (syclik) [inline stan/math/prim/scal/prob/cauchy_log.hpp]: I did go through and fix these.

**Comment by seantalts** [comment]
I'll take a look through tomorrow, but feel free to not wait on me. I'll look either way to start getting familiar with the goings-on and eventually hope to take some pull request load off you guys.

**Comment by syclik** [comment]
Thanks. No rush.

I did the work, Bob approved it a while ago, but I updated a few things
recently. I just wanted another set of eyes on it before it got in.

**Comment by seantalts** [comment]
PS when this is merged in, we should have some kind of party - it's not every day you get to delete 10,000 lines of code!

**Comment by syclik** [comment]
=). In general, I'm with you and we should celebrate when we kill a lot of
code! In this instance, I had to duplicate code for the _log to _lpdf /
_lpmf change.

It would have been in a single pull request, but this started when GitHub
didn't do diffs so well.

**Comment by seantalts** [comment]
Changes seem good to me, though tbh I mostly looked for general replacement patterns and didn't read all 10000 lines of diff.

**Comment by syclik** [comment]
That's good. Feel free to merge when you're ready.

**Pull request merged**